perception/map_based_prediction/src/path_generator.cpp

```diff
@@ -0,0 +1,346 @@
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "path_generator.hpp"
+
+#include <interpolation/spline_interpolation.hpp>
+#include <tier4_autoware_utils/tier4_autoware_utils.hpp>
+
+namespace map_based_prediction
+{
+PathGenerator::PathGenerator(const double time_horizon, const double sampling_time_interval)
+: time_horizon_(time_horizon), sampling_time_interval_(sampling_time_interval)
+{
+}
+
+PredictedPath PathGenerator::generatePathForNonVehicleObject(
+  const std_msgs::msg::Header & origin_header, const TrackedObject & object)
+{
+  return generateStraightPath(origin_header, object);
+}
+
+PredictedPath PathGenerator::generatePathForLowSpeedVehicle(
+  const std_msgs::msg::Header & origin_header, const TrackedObject & object) const
+{
+  PredictedPath path;
+  path.time_step = rclcpp::Duration::from_seconds(sampling_time_interval_);
+  const double ep = 0.001;
+  for (double dt = 0.0; dt < time_horizon_ + ep; dt += sampling_time_interval_) {
+    path.path.push_back(object.kinematics.pose_with_covariance.pose);
+  }
+
+  return path;
+}
+
+PredictedPath PathGenerator::generatePathForOffLaneVehicle(
+  const std_msgs::msg::Header & origin_header, const TrackedObject & object)
+{
+  return generateStraightPath(origin_header, object);
+}
+
+PredictedPath PathGenerator::generatePathForOnLaneVehicle(
+  const std_msgs::msg::Header & origin_header, const TrackedObject & object,
+  const PosePath & ref_paths)
+{
+  if (ref_paths.size() < 2) {
+    const PredictedPath empty_path;
+    return empty_path;
+  }
+
+  return generatePolynomialPath(origin_header, object, ref_paths);
+}
+
+PredictedPath PathGenerator::generateStraightPath(
+  const std_msgs::msg::Header & origin_header, const TrackedObject & object) const
+{
+  const auto & object_pose = object.kinematics.pose_with_covariance.pose;
+  const auto & object_twist = object.kinematics.twist_with_covariance.twist;
+
+  PredictedPath path;
+  path.time_step = rclcpp::Duration::from_seconds(sampling_time_interval_);
+  const double ep = 0.001;
+  for (double dt = 0.0; dt < time_horizon_ + ep; dt += sampling_time_interval_) {
+    geometry_msgs::msg::Pose object_frame_pose;
+    geometry_msgs::msg::Pose world_frame_pose;
+    object_frame_pose.position.x = object_twist.linear.x * dt;
+    object_frame_pose.position.y = object_twist.linear.y * dt;
+    tf2::Quaternion quat;
+    quat.setRPY(0.0, 0.0, 0.0);
+    object_frame_pose.orientation = tf2::toMsg(quat);
+    tf2::Transform tf_object2future;
+    tf2::Transform tf_world2object;
+    tf2::Transform tf_world2future;
+
+    tf2::fromMsg(object_pose, tf_world2object);
+    tf2::fromMsg(object_frame_pose, tf_object2future);
+    tf_world2future = tf_world2object * tf_object2future;
+    tf2::toMsg(tf_world2future, world_frame_pose);
```

Comment:
For me, much easier to understand if the code is like this
```
const auto future_obj_pose = tier4_autoware_utils::calcOffsetPose(object_pose, obj_twist.linear.x * dt, obj_twist.linear.y * dt, 0.0);
path.path.push_back(future_obj_pose); 
```

Reply:
done